query-builder/src/test/java/com/datastax/oss/driver/api/querybuilder/select/SelectOrderingTest.java

```diff
@@ -46,6 +47,103 @@ public void should_generate_ordering_clauses() {
         .hasCql("SELECT * FROM foo WHERE k=1 ORDER BY c1 ASC,c2 DESC");
   }
 
+  @Test
+  public void should_support_nested_sub_conditions() {
+    assertThat(
+            selectFrom("foo")
+                .all()
+                .where(Relation.column("k").isEqualTo(literal(1)))
+                .and()
+                .where(
+                    subCondition()
+                        .where(Relation.column("l").isEqualTo(literal(2)))
+                        .where(
+                            subCondition()
+                                .where(Relation.column("m").isEqualTo(literal(3)))
+                                .or()
+                                .where(Relation.column("x").isEqualTo(literal(4)))))
+                .orderBy("c1", ASC))
+        .hasCql("SELECT * FROM foo WHERE k=1 AND (l=2 AND (m=3 OR x=4)) ORDER BY c1 ASC");
+  }
+
+  @Test
+  public void should_generate_criteria_alternative() {
+    assertThat(
+            selectFrom("foo")
+                .all()
+                .where(Relation.column("k").isEqualTo(literal(1)))
+                .or()
+                .where(Relation.column("l").isEqualTo(literal(2)))
+                .orderBy("c1", ASC)
+                .orderBy("c2", DESC))
+        .hasCql("SELECT * FROM foo WHERE k=1 OR l=2 ORDER BY c1 ASC,c2 DESC");
+  }
+
+  @Test
+  public void should_support_sub_condition() {
+    assertThat(
+            selectFrom("foo")
+                .all()
+                .where(Relation.column("k").isEqualTo(literal(1)))
+                .and()
+                .where(
+                    subCondition()
+                        .where(Relation.column("l").isEqualTo(literal(2)))
+                        .or()
+                        .where(Relation.column("m").isEqualTo(literal(3))))
+                .orderBy("c1", ASC)
+                .orderBy("c2", DESC))
+        .hasCql("SELECT * FROM foo WHERE k=1 AND (l=2 OR m=3) ORDER BY c1 ASC,c2 DESC");
```

Comment:
If I understand correctly, this a 'recursive' building.
So we can have 

`SELECT * FROM foo WHERE k=1 AND (l=2 OR m=3 OR (q=4 AND p=5 )) ORDER BY c1 ASC,c2 DESC"`

Or other complex nested logical query?

Reply:
Yes, you are correct. I have made one of the queries more complex.